### DISCUSSION TITLE
Exporting CellProfiler area and shape measurements in non-pixel units

### DISCUSSION THREAD

**Discussion by bethac07** [CellProfiler/CellProfiler#4887] [General]
Copying and subsetting from Slack:

> @emiglietta
Question from the workshop: is it possible to export CellProfiler distance measurements in um instead of pixels? is there any place to input X,Y scale when loading pictures or can it be retrieved from image metadata?
Finally if this doesn't exist, could it be a mini project for updating a CP module? Thanks!
@bethac07 
Short version - not easily currently
Though you can add CalculateMath modules to the pipeline to add it
I think in general we have not assumed we could reliably get it from the Metadata, so decided we wouldn't get into the game at all
@AnneCarpenter 
Anne Carpenter
Ha! this was originally a feature in CP in the earliest days, but we took it out because it’s hard to ensure that all relevant measurements (esp in plugins) included the conversion factor. We figured folks could do the math later.

**Comment by bethac07** [comment]
In general, Anne provides the takehome point. 

Here are the modules where I think potentially we might consider a distance metric that's non-pixel based, in approximate order of "how often they are requested, and/or how useful I think they'd be". Realistically, it's the first 1-to-maybe-3. 
- MeasureObjectSizeShape
- MeasureObjectNeighbors
- Added from Anne's comments - Resize & ResizeObjects
- TrackObjects
- MeasureImageAreaOccupied
- DistanceTransform (now we're getting into the ones where this is a stretch)
- EnhanceOrSuppressFeatures

TLDR, I don't think it ultimately is worth the pain to do this, but if we're going to do ANYTHING but "always work in pixel units" in, for my money, we should only do this in measurement modules where we can specify if this is a measurement in (pixels, nm, µm, etc) as part of the actual measurement name. Otherwise, it's too hard to compare across workflows because you can't be sure what a given measurement means.

Here are my thoughts of possible implementations and possible problems I see. Feel free to let me know what I missed.
- In one or a few select modules (like MOSS), we allow the users to optionally hardcode a pixel size conversion factor. 
  - Pro: self-contained, easy
  - Cons: way too easy to forget when adjusting (binning, magnification, etc) to adjust here. Also, what do we do when there is more than one value per data set (which there SHOULD NOT BE, but users are going to user)? 
- Same, but hardcoded in NamesAndTypes (as an option, default off) and then inherited. Individual modules have to enable unit conversion.
  - Pros and cons are similar to above, but at least it's set at the beginning, where we already DO set pixel sizes for 3D. This is the version I could most easily be talked into (but I still don't love it)
- Try to read the Metadata on the pixel size, fall back on pixels if not (presumably set in M or N&T)
  - Pro: easiest for the user, replicates what they expect from ie Fiji
  - Con: I currently feel pretty confident we should not do this but I'm willing to be talked out of it, but it would take a bit. It makes measurement names non-deterministic from the pipeline, to start. I don't feel like we want CellProfiler returning different values in an identical pipeline based on whether or not the metadata extraction succeeds. What happens when we extract from some or not others? What happens when the extraction is wrong (aka something gets exported in DPI rather than, say, nm/µm, which we see not-infrequently)?
    - I think the version I could second-most-easily be talked into would be one where we do this but CellProfiler just fails when you've told it to use a calibration and it cannot find one (and an identical one at that) for 100% of images. I still don't like the non-deterministic measurement thing but I might not die.

**Comment by AnneCarpenter** [comment]
I recall another complication - any modules that change the size of the image also need to carry along the size adjustment. So anything that downsamples or upsamples the image would need attention as well.
